File: mobile/backend-fixes/ClubAuthorizationService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using GatherGrove.Infrastructure.Data;

namespace GatherGrove.Infrastructure.Services;

/// <summary>
/// Service for handling club-level authorization and feature access control
/// </summary>
public class ClubAuthorizationService
{
    private readonly GatherGroveDbContext _context;
    private readonly ILogger<ClubAuthorizationService> _logger;

    // Define tier-based feature access rules
    private static readonly Dictionary<string, HashSet<string>> TierFeatureAccess = new()
    {
        ["Basic"] = new HashSet<string>
        {
            "BasicEventManagement",
            "MemberDirectory", 
            "EventRSVP",
            "BasicReporting"
        },
        ["Growth"] = new HashSet<string>
        {
            "BasicEventManagement",
            "MemberDirectory",
            "EventRSVP", 
            "BasicReporting",
            "AdvancedEventManagement",
            "MemberCommunication",
            "EventAnalytics"
        },
        ["Unlimited"] = new HashSet<string>
        {
            "BasicEventManagement",
            "MemberDirectory",
            "EventRSVP",
            "BasicReporting",
            "AdvancedEventManagement", 
            "MemberCommunication",
            "EventAnalytics",
            "EventEngagementAnalytics",
            "MemberEngagementInsights",
            "EventPerformanceAnalysis",
            "EngagementTrends",
            "EventRecommendations",
            "ROIMetrics"
        }
    };

    public ClubAuthorizationService(GatherGroveDbContext context, ILogger<ClubAuthorizationService> logger)
    {
        _context = context;
        _logger = logger;
    }

    /// <summary>
    /// Validate if a user has access to a specific club
    /// </summary>
    public async Task<bool> ValidateClubAccessAsync(int clubId, int userId)
    {
        try
        {
            // Check if user is a club admin (owner/admin role)
            var isAdmin = await _context.ClubAdmins
                .AnyAsync(ca => ca.ClubId == clubId && ca.UserId == userId);

            if (isAdmin)
                return true;

            // Check if user is an active member
            var isMember = await _context.Members
                .AnyAsync(m => m.ClubId == clubId && m.Status == "Active" && 
                              m.Email == _context.Users.Where(u => u.Id == userId).Select(u => u.Email).FirstOrDefault());

            return isMember;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error validating club access for user {UserId} in club {ClubId}", userId, clubId);
            return false;
        }
    }

    /// <summary>
    /// Check if a club has access to a specific feature based on tier
    /// </summary>
    public async Task<bool> HasFeatureAccess(int clubId, string featureName)
    {
        try
        {
            if (string.IsNullOrEmpty(featureName))
                return false;

            var club = await _context.Clubs.FirstOrDefaultAsync(c => c.Id == clubId);
            if (club == null)
                return false;

            // Check for expired subscriptions
            if (club.TrialExpiresAt.HasValue && club.TrialExpiresAt.Value < DateTime.UtcNow)
            {
                _logger.LogWarning("Club {ClubId} trial has expired, denying feature access", clubId);
                return false;
            }

            var clubTier = club.Tier ?? "Basic";

            // Check if the tier has access to the requested feature
            if (TierFeatureAccess.ContainsKey(clubTier))
            {
                var hasAccess = TierFeatureAccess[clubTier].Contains(featureName);
                
                if (!hasAccess)
                {
                    _logger.LogWarning("Club {ClubId} with tier {Tier} denied access to feature {Feature}", 
                        clubId, clubTier, featureName);
                }

                return hasAccess;
            }

            _logger.LogWarning("Unknown club tier {Tier} for club {ClubId}", clubTier, clubId);
            return false;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error checking feature access for club {ClubId}, feature {Feature}", clubId, featureName);
            return false;
        }
    }

    /// <summary>
    /// Get user's role within a club
    /// </summary>
    public async Task<string> GetUserRoleInClub(int clubId, int userId)
    {
        try
        {
            // Check if user is owner (created the club)
            var club = await _context.Clubs.FirstOrDefaultAsync(c => c.Id == clubId && c.CreatedByUserId == userId);
            if (club != null)
                return "Owner";

            // Check if user is admin
            var isAdmin = await _context.ClubAdmins.AnyAsync(ca => ca.ClubId == clubId && ca.UserId == userId);
            if (isAdmin)
                return "Admin";

            // Check if user is member
            var userEmail = await _context.Users.Where(u => u.Id == userId).Select(u => u.Email).FirstOrDefaultAsync();
            var isMember = await _context.Members.AnyAsync(m => m.ClubId == clubId && m.Email == userEmail && m.Status == "Active");
            if (isMember)
                return "Member";

            return "None";
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user role for user {UserId} in club {ClubId}", userId, clubId);
            return "None";
        }
    }

    /// <summary>
    /// Check if user has administrative permissions in a club
    /// </summary>
    public async Task<bool> HasAdministrativeAccess(int clubId, int userId)
    {
        var role = await GetUserRoleInClub(clubId, userId);
        return role == "Owner" || role == "Admin";
    }
}

File: mobile/src/services/clubAuthorizationService.ts
/**
 * Club Authorization Service
 * Handles club-level authorization and tier-based feature access control
 */

import { authService } from './authService';

export interface ClubTierFeatures {
  [tier: string]: string[];
}

export interface Club {
  id: number;
  name: string;
  tier: string;
  createdByUserId: number;
  trialExpiresAt: string | null;
}

class ClubAuthorizationService {
  // Define tier-based feature access rules matching backend implementation
  private static readonly TIER_FEATURE_ACCESS: ClubTierFeatures = {
    'Basic': [
      'BasicEventManagement',
      'MemberDirectory', 
      'EventRSVP',
      'BasicReporting'
    ],
    'Growth': [
      'BasicEventManagement',
      'MemberDirectory',
      'EventRSVP', 
      'BasicReporting',
      'AdvancedEventManagement',
      'MemberCommunication',
      'EventAnalytics'
    ],
    'Unlimited': [
      'BasicEventManagement',
      'MemberDirectory',
      'EventRSVP',
      'BasicReporting',
      'AdvancedEventManagement', 
      'MemberCommunication',
      'EventAnalytics',
      'EventEngagementAnalytics',
      'MemberEngagementInsights',
      'EventPerformanceAnalysis',
      'EngagementTrends',
      'EventRecommendations',
      'ROIMetrics'
    ],
    'Expand': [
      'BasicEventManagement',
      'MemberDirectory',
      'EventRSVP',
      'BasicReporting',
      'AdvancedEventManagement', 
      'MemberCommunication',
      'EventAnalytics',
      'EventEngagementAnalytics',
      'MemberEngagementInsights',
      'EventPerformanceAnalysis',
      'EngagementTrends',
      'EventRecommendations',
      'ROIMetrics'
    ]
  };

  /**
   * Validate if a user has access to a specific club
   */
  async validateClubAccess(clubId: number, userId: number): Promise<boolean> {
    try {
      // Check if user is a club admin (owner/admin role)
      const isAdmin = await this.isClubAdmin(clubId, userId);
      if (isAdmin) {
        return true;
      }

      // Check if user is an active member
      const isMember = await this.isActiveMember(clubId, userId);
      return isMember;
    } catch (error) {
      return false;
    }
  }

  /**
   * Check if a club has access to a specific feature based on tier
   */
  async hasFeatureAccess(clubId: number, featureName: string): Promise<boolean> {
    try {
      if (!featureName) {
        return false;
      }

      const club = await this.getClubById(clubId);
      if (!club) {
        return false;
      }

      // DATE-02 fix: Use explicit timestamp comparison for timezone consistency
      if (club.trialExpiresAt) {
        const expiresAt = new Date(club.trialExpiresAt).getTime();
        if (expiresAt < Date.now()) {
          return false;
        }
      }

      const clubTier = club.tier || 'Basic';

      // Check if the tier has access to the requested feature
      if (ClubAuthorizationService.TIER_FEATURE_ACCESS[clubTier]) {
        const hasAccess = ClubAuthorizationService.TIER_FEATURE_ACCESS[clubTier].includes(featureName);

        if (!hasAccess) {
          /* Feature not available for this tier */
        }

        return hasAccess;
      }

      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Get user's role within a club
   */
  async getUserRoleInClub(clubId: number, userId: number): Promise<string> {
    try {
      // Check if user is owner (created the club)
      const club = await this.getClubById(clubId);
      if (club && club.createdByUserId === userId) {
        return 'Owner';
      }

      // Check if user is admin
      const isAdmin = await this.isClubAdmin(clubId, userId);
      if (isAdmin) {
        return 'Admin';
      }

      // Check if user is member
      const isMember = await this.isActiveMember(clubId, userId);
      if (isMember) {
        return 'Member';
      }

      return 'None';
    } catch (error) {
      return 'None';
    }
  }

  /**
   * Check if user has administrative permissions in a club
   */
  async hasAdministrativeAccess(clubId: number, userId: number): Promise<boolean> {
    const role = await this.getUserRoleInClub(clubId, userId);
    return role === 'Owner' || role === 'Admin';
  }

  /**
   * Get club tier for a specific club
   */
  async getClubTier(clubId: number): Promise<string> {
    try {
      const club = await this.getClubById(clubId);
      return club?.tier || 'Basic';
    } catch (error) {
      return 'Basic';
    }
  }

  // Private helper methods

  private async getClubById(clubId: number): Promise<Club> {
    // Mock implementation - in real app, this would fetch from API
    return {
      id: clubId,
      name: 'Sample Club',
      tier: 'Expand', // Can be 'Basic', 'Growth', 'Expand', or legacy 'Unlimited'
      createdByUserId: 1,
      trialExpiresAt: null // or a future date
    };
  }

  private async isClubAdmin(clubId: number, userId: number): Promise<boolean> {
    // MOCK-03 fix: Use actual user data instead of hardcoded test values
    // Check if the current authenticated user has admin role for this club
    const currentUser = await authService.getCurrentUser();
    if (!currentUser?.user) {
      return false;
    }

    // User is admin if they have the Admin role and belong to this club
    const isAdmin = currentUser.user.role === 'Admin' || currentUser.user.role === 'Owner';
    const belongsToClub = currentUser.user.clubId === clubId;
    const isCorrectUser = currentUser.user.userId === userId;

    return isAdmin && belongsToClub && isCorrectUser;
  }

  private async isActiveMember(_clubId: number, userId: number): Promise<boolean> {
    // Mock implementation - in real app, this would check member status
    const currentUser = await authService.getCurrentUser();
    
    // For testing, assume user is active member if they have access to the club
    return currentUser?.user?.userId === userId || false;
  }

  /**
   * Get all features available to a specific tier
   */
  getTierFeatures(tier: string): string[] {
    return ClubAuthorizationService.TIER_FEATURE_ACCESS[tier] || [];
  }

  /**
   * Check if a specific tier has access to a feature
   */
  doesTierHaveFeature(tier: string, featureName: string): boolean {
    const tierFeatures = this.getTierFeatures(tier);
    return tierFeatures.includes(featureName);
  }
}

export const clubAuthorizationService = new ClubAuthorizationService();
